homeassistant/components/fujitsu_hvac/climate.py

```diff
@@ -0,0 +1,202 @@
+        return
+
+    devices = await api.async_get_devices()
+    for dev in devices:
+        if dev.device_serial_number == entry.data[CONF_DEVICE]:
+            await dev.async_update()
+            async_add_entities([FujitsuHVACDevice(dev)])
+            return
+
+
+class FujitsuHVACDevice(ClimateEntity):
+    """Represent a Fujitsu HVAC device."""
+
+    _attr_temperature_unit = UnitOfTemperature.CELSIUS
+    _attr_precision = 0.5
+    _attr_target_temperature_step = 0.5
+    _attr_has_entity_name = True
+    _attr_name = None
+
+    def __init__(self, dev: FujitsuHVAC) -> None:
+        """Set up static attributes."""
+        self._dev = dev
+
+        self._attr_unique_id = f"{DOMAIN}_{self._dev.device_serial_number}"
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, self._dev.device_serial_number)},
+            name=self._dev.device_name,
+            manufacturer="Fujitsu",
+            model=self._dev.property_values["model_name"],
+            serial_number=self._dev.device_serial_number,
+            sw_version=self._dev.property_values["mcu_firmware_version"],
+        )
+        self.update_attributes()
+
+    async def async_update(self) -> None:
+        """Gather data from API and update our attributes."""
+        try:
+            if self._dev.ayla_api.token_expiring_soon:
+                await self._dev.ayla_api.refresh_auth()
+        except AylaAuthError:
+            await self._dev.ayla_api.async_sign_in()
+
+        await self._dev.async_update()
+        self.update_attributes()
+
+    def update_attributes(self) -> None:
+        """Update all dynamic attributes."""
+        self._attr_supported_features = self._get_supported_features()
+        self._attr_current_temperature = self._dev.sensed_temp
+        self._attr_fan_mode = self._get_fan_mode()
+        self._attr_fan_modes = self._get_fan_modes()
+        self._attr_hvac_mode = self._get_hvac_mode()
+        self._attr_hvac_modes = self._get_hvac_modes()
+        self._attr_swing_mode = self._get_swing_mode()
+        self._attr_swing_modes = self._get_swing_modes()
+        self._attr_target_temperature = self._get_target_temperature()
+        (
+            self._attr_min_temp,
+            self._attr_max_temp,
+        ) = self._dev.temperature_range
+
+    def _get_fan_mode(self) -> str | None:
+        with suppress(KeyError):
+            return FAN_MODE_MAP.inverse[self._dev.fan_speed]
+
+        return None
+
+    def _get_fan_modes(self) -> list[str] | None:
+        ret = [
+            FAN_MODE_MAP.inverse[mode]
+            for mode in self._dev.supported_fan_speeds
+            if mode in FAN_MODE_MAP.inverse
+        ]
+
+        if len(ret) > 0:
+            return ret
+
+        return None
+
+    def _get_hvac_mode(self) -> HVACMode | None:
+        with suppress(KeyError):
+            return HVAC_MODE_MAP.inverse[self._dev.op_mode]
+
+        return None
+
+    def _get_hvac_modes(self) -> list[HVACMode]:
```

Comment:
regarding your question this is what I meant:
```suggestion
    @property
    def hvac_modes(self) -> list[HVACMode]:
```
by using the built in properties, you don't need to define a method that you then call to assign to an attribute, but can just put the code inside the property definition

Reply:
I tried doing that but it raised an exception from here : [raise TypeError(f"Can't override {attr_name} in subclass")](https://github.com/crevetor/ha-core/blob/9d4d423975c7e5f5cbb6f575e5380f5f7b3fc886/homeassistant/helpers/entity.py#L386) so it seems you can't use property decorated methods for those attributes (IIRC I also tried def _attr_... which didn't work either).

Reply:
Actually, after looking at it closer it seems that this exception is only raised if _attr_... is a property decorated method, I could override the methods defined in ClimateEntity but I think I prefer keeping it this way because this way we use the way ClimateEntity handles those properties (specifically ClimateEntity decorates these properties with cached_property.

Reply:
I really think it would make your code a lot cleaner (compare it to ecobee's climate implementation https://github.com/home-assistant/core/blob/dev/homeassistant/components/ecobee/climate.py). Not sure about the `@cached_property` but maybe you can just redecorate yours the same way.